File: python/Burst Balloons.py
class Solution(object):
    def maxCoins(self, nums):
        nums = [1] + nums + [1]
        n = len(nums)
        dp = [[0] * n for _ in xrange(n)]
        for l in xrange(2, n):
            for left in xrange(0, n - l):
                right = left + l
                for k in xrange(left + 1, right):
                    dp[left][right] = max(dp[left][right],                         nums[left] * nums[k] * nums[right] + dp[left][k] + dp[k][right])
        
        return dp[0][n - 1]

File: cpp/Burst Balloons.cpp
class Solution {
public:
    int maxCoins(vector<int>& nums) {
        int dp[500+10][500+10] = {0};
        nums.insert(nums.begin(), 1);
        nums.push_back(1);
        
        for (int len = 2; len < nums.size(); len++) {
            for (int left = 0; left + len < nums.size(); left++) {
                int right = left + len;
                for (int k = left + 1; k < right; k++) {
                    int tmp = nums[left] * nums[k] * nums[right] + dp[left][k] + dp[k][right];
                    dp[left][right] = max(dp[left][right], tmp);
                }
            }
        }
        
        return dp[0][nums.size() - 1];
    }
};
